alws/crud/distribution.py

```diff
@@ -152,38 +179,44 @@ async def modify_distribution(build_id: int, distribution: str, db: Session,
             models.Build.id.__eq__(build_id)
         ).options(
             selectinload(models.Build.tasks).selectinload(
-                models.BuildTask.artifacts),
-            selectinload(models.Build.tasks).selectinload(
-                models.BuildTask.rpm_module)
+                models.BuildTask.rpm_module),
+            selectinload(models.Build.repos)
         ))
         db_build = db_build.scalars().first()
 
-        pulp_client = PulpClient(settings.pulp_host, settings.pulp_user,
-                                 settings.pulp_password)
-
-        if modification == 'add':
+        if modification == 'add' and not force:
             if db_build in db_distro.builds:
                 error_msg = f'Packages of build {build_id} have already been' \
                             f' added to {distribution} distribution'
                 raise DistributionError(error_msg)
-            db_distro.builds.append(db_build)
-        if modification == 'remove':
+        if modification == 'remove' and not force:
             if db_build not in db_distro.builds:
                 error_msg = f'Packages of build {build_id} cannot be removed ' \
                             f'from {distribution} distribution ' \
                             f'as they are not added there'
                 raise DistributionError(error_msg)
-            remove_query = models.Build.id.__eq__(build_id)
-            await db.execute(
-                delete(models.DistributionBuilds).where(remove_query)
-            )
 
-        await db.commit()
-    await db.refresh(db_distro)
-    modify = await prepare_repo_modify_dict(db_build, db_distro)
+    pulp_client = PulpClient(settings.pulp_host, settings.pulp_user,
+                             settings.pulp_password)
+    modify = await prepare_repo_modify_dict(db_build, db_distro, pulp_client)
+    tasks = []
     for key, value in modify.items():
         if modification == 'add':
-            await pulp_client.modify_repository(add=value, repo_to=key)
+            tasks.append(pulp_client.modify_repository(add=value, repo_to=key))
         else:
-            await pulp_client.modify_repository(remove=value, repo_to=key)
-        await pulp_client.create_rpm_publication(key)
+            tasks.append(pulp_client.modify_repository(
+                remove=value, repo_to=key))
+    await asyncio.gather(*tasks)
+
+    if modification == 'add':
+        db_distro.builds.append(db_build)
+    else:
+        remove_query = models.Build.id.__eq__(build_id)
+        await db.execute(
+            delete(models.DistributionBuilds).where(remove_query)
+        )
+    db.add(db_distro)
+    try:
+        await db.commit()
```

Comment:
If there will be exception during `db.commit`, does session rollback working? or it's should be called manually?

Reply:
Ideally, we should do it in the separate context manager)

Reply:
Ideally...